hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java

```diff
@@ -142,34 +145,52 @@ private Map<String, DbField> readDbFields(
                 OPTION_EXTERNAL_DATASTORE_REF + " must be set"
         );
         DataSource dataSource = createDataStore(nodeEngine, externalDataStoreRef);
-        try (
-                Connection connection = dataSource.getConnection();
-                Statement statement = connection.createStatement()
-        ) {
-            Set<String> pkColumns = readPrimaryKeyColumns(externalTableName, connection);
-
-            boolean hasResultSet = statement.execute("SELECT * FROM " + externalTableName + " LIMIT 0");
-            if (!hasResultSet) {
-                throw new IllegalStateException("Could not resolve fields for table " + externalTableName);
-            }
-            ResultSet rs = statement.getResultSet();
-            ResultSetMetaData metaData = rs.getMetaData();
+        try (Connection connection = dataSource.getConnection()) {
+
+            DatabaseMetaData databaseMetaData = connection.getMetaData();
+
+            Set<String> pkColumns = readPrimaryKeyColumns(externalTableName, databaseMetaData);
+
+            return readColumns(externalTableName, databaseMetaData, pkColumns);
 
-            Map<String, DbField> fields = new LinkedHashMap<>();
-            for (int i = 1; i <= metaData.getColumnCount(); i++) {
-                String columnName = metaData.getColumnName(i);
+        } catch (Exception e) {
+            throw new HazelcastException("Could not execute readDbFields for table " + externalTableName, e);
+        } finally {
+            closeDataSource(dataSource);
+        }
+    }
+
+    private static Map<String, DbField> readColumns(String externalTableName, DatabaseMetaData databaseMetaData,
```

Comment:
The `readPrimaryKeyColumns` is called first, so should be reordered with `readColumns`

Reply:
Ok